File: client/validation.lua
local function normalizeActions(actions)
    if type(actions) ~= 'table' then return nil end
    local normalized, seen = {}, {}
    for _, action in ipairs(actions) do
        if #normalized >= Config.Limits.actions then break end
        if type(action) == 'table' then
            local id = SyncNotify.Text(action.id, 32)
            local label = SyncNotify.Text(action.label, Config.Limits.actionLabel)
            if id and id ~= '' and label and label ~= '' and not seen[id] then
                seen[id] = true
                normalized[#normalized + 1] = { id = id, label = label }
            end
        end
    end
    return #normalized > 0 and normalized or nil
end

function SyncNotify.Normalize(input, patch)
    if type(input) == 'string' then input = { message = input } end
    if type(input) ~= 'table' then return nil, 'options must be a table' end

    local result = {}
    local kind = SyncNotify.Types[input.type] and input.type or (patch and input.type or 'info')
    if kind then result.type = kind end
    if input.id ~= nil then result.id = SyncNotify.Text(input.id, 64) end
    if input.title ~= nil then result.title = SyncNotify.Text(input.title, Config.Limits.title, '') end
    if input.message ~= nil then result.message = SyncNotify.Text(input.message, Config.Limits.message, '') end
    if not patch and (not result.message or result.message == '') then return nil, 'message is required' end

    if input.icon ~= nil then
        local icon = SyncNotify.Text(input.icon, 48)
        result.icon = SyncNotify.BuiltinIcons[icon] and icon or 'bell'
    end
    if input.duration ~= nil then result.duration = SyncNotify.Clamp(input.duration, Config.Limits.minDuration, Config.Limits.maxDuration) end
    if not patch and result.duration == nil then result.duration = Config.Duration end
    if input.persistent ~= nil then result.persistent = input.persistent == true end
    if input.priority ~= nil then result.priority = math.floor(SyncNotify.Clamp(input.priority, 0, 3) or 0) end
    if not patch and result.priority == nil then result.priority = 0 end
    if input.position ~= nil then result.position = SyncNotify.Positions[input.position] and input.position or Config.Position end
    if not patch and result.position == nil then result.position = Config.Position end
    if input.sound ~= nil then result.sound = type(input.sound) == 'string' and SyncNotify.Text(input.sound, 32) or input.sound == true end
    if input.progress ~= nil then
        result.progress = type(input.progress) == 'number' and SyncNotify.Clamp(input.progress, 0, 100) or input.progress == true
    end
    if input.progressStyle ~= nil then result.progressStyle = SyncNotify.ProgressStyles[input.progressStyle] and input.progressStyle or 'rail' end
    if input.mode ~= nil then result.mode = SyncNotify.Modes[input.mode] and input.mode or Config.Mode end
    if not patch and result.mode == nil then result.mode = SyncNotify.Modes[Config.Mode] and Config.Mode or 'auto' end
    local defaultDesign = SyncNotify.Designs[Config.Design] and Config.Design or 'floating'
    if input.design ~= nil then result.design = SyncNotify.Designs[input.design] and input.design or defaultDesign end
    if not patch and result.design == nil then result.design = defaultDesign end
    if input.duplicateMode ~= nil then result.duplicateMode = SyncNotify.DuplicateModes[input.duplicateMode] and input.duplicateMode or 'allow' end
    if input.theme ~= nil then result.theme = SyncNotify.Text(input.theme, 32) end
    if input.actions ~= nil then result.actions = normalizeActions(input.actions) end
    if input.metadata ~= nil then result.metadata = SyncNotify.SafeCopy(input.metadata, Config.Limits.metadataDepth) end
    return result
end

function SyncNotify.NormalizeTheme(definition)
    if type(definition) ~= 'table' then return nil end
    local theme = {
        accent = SyncNotify.SafeColor(definition.accent),
        icon = SyncNotify.BuiltinIcons[definition.icon] and definition.icon or nil,
        sound = SyncNotify.Text(definition.sound, 32)
    }
    if not theme.accent and not theme.icon and not theme.sound then return nil end
    return theme
end

File: web/src/utils/validation.ts
import type { NotificationOptions, NotificationItem, NotificationPosition, NotificationType, DuplicateMode, ProgressStyle, NotificationMode, ResolvedNotificationMode, NotificationDesign } from '../types/notification'

const types = new Set<NotificationType>(['success', 'error', 'warning', 'info', 'dispatch', 'custom'])
const positions = new Set<NotificationPosition>(['top-left', 'top-center', 'top-right', 'middle-left', 'middle-right', 'bottom-left', 'bottom-center', 'bottom-right'])
const duplicates = new Set<DuplicateMode>(['allow', 'replace', 'increment', 'refresh'])
const progressStyles = new Set<ProgressStyle>(['rail', 'minimal', 'none'])
const modes = new Set<NotificationMode>(['auto', 'micro', 'full'])
const designs = new Set<NotificationDesign>(['split', 'floating'])
const trim = (value: unknown, max: number) => typeof value === 'string' ? Array.from(value).filter(character => character.charCodeAt(0) >= 32).join('').slice(0, max) : undefined
const clamp = (value: unknown, min: number, max: number, fallback: number) => typeof value === 'number' && Number.isFinite(value) ? Math.min(max, Math.max(min, value)) : fallback

export function resolveNotificationMode(input: Pick<NotificationOptions, 'mode' | 'title' | 'actions' | 'type' | 'priority'>): ResolvedNotificationMode {
  if (input.mode === 'micro' || input.mode === 'full') return input.mode
  return !input.title && !input.actions?.length && input.type !== 'dispatch' && (input.priority ?? 0) <= 1 ? 'micro' : 'full'
}

export function normalizeDesign(value: unknown, fallback: NotificationDesign = 'floating'): NotificationDesign {
  return designs.has(value as NotificationDesign) ? value as NotificationDesign : fallback
}

export function normalizeNotification(input: NotificationOptions, now = Date.now(), defaultDesign: NotificationDesign = 'floating'): NotificationItem | null {
  const message = trim(input?.message, 500)
  if (!message) return null
  const duration = clamp(input.duration, 1000, 60000, 5000)
  const handle = trim(input.handle, 80) || `web:${now}:${Math.random().toString(36).slice(2, 8)}`
  const actions = Array.isArray(input.actions) ? input.actions.slice(0, 3).flatMap(action => {
    const id = trim(action?.id, 32); const label = trim(action?.label, 32)
    return id && label ? [{ id, label }] : []
  }).filter((action, index, all) => all.findIndex(value => value.id === action.id) === index) : undefined
  const mode = modes.has(input.mode as NotificationMode) ? input.mode! : 'auto'
  const normalized = {
    handle, id: trim(input.id, 64), type: types.has(input.type as NotificationType) ? input.type! : 'info',
    theme: trim(input.theme, 32), title: trim(input.title, 80), message, icon: trim(input.icon, 48),
    duration, persistent: input.persistent === true, priority: Math.floor(clamp(input.priority, 0, 3, 0)),
    position: positions.has(input.position as NotificationPosition) ? input.position! : 'top-right',
    sound: input.sound === true ? true : typeof input.sound === 'string' ? trim(input.sound, 32) : input.sound === false ? false : undefined,
    progress: typeof input.progress === 'number' ? clamp(input.progress, 0, 100, 0) : input.progress === true,
    progressStyle: progressStyles.has(input.progressStyle as ProgressStyle) ? input.progressStyle! : 'rail',
    duplicateMode: duplicates.has(input.duplicateMode as DuplicateMode) ? input.duplicateMode! : 'allow',
    actions: actions?.length ? actions : undefined, metadata: input.metadata, mode, design: normalizeDesign(input.design, defaultDesign),
    createdAt: now, startedAt: now, remaining: duration, paused: false, count: 1, phase: 'entering' as const, actionLocked: false
  }
  return { ...normalized, resolvedMode: resolveNotificationMode(normalized) }
}
